compiler-plugin/src/main/java/io/ballerina/lib/np/compilerplugin/ConstantExpressionVisitor.java

```diff
@@ -0,0 +1,375 @@
+        addConstantReferencesDiagnostics(implicitAnonymousFunctionExpressionNode.toSourceCode());
+    }
+
+    @Override
+    public void visit(ImplicitNewExpressionNode implicitNewExpressionNode) {
+        addConstantReferencesDiagnostics(implicitNewExpressionNode.toSourceCode());
+    }
+
+    @Override
+    public void visit(IndexedExpressionNode indexedExpressionNode) {
+        visitSyntaxNode(indexedExpressionNode.containerExpression());
+        indexedExpressionNode.keyExpression().forEach(node -> visitSyntaxNode(node));
+    }
+
+    @Override
+    public void visit(IntersectionTypeDescriptorNode intersectionTypeDescriptorNode) {
+        addConstantReferencesDiagnostics(intersectionTypeDescriptorNode.toSourceCode());
+    }
+
+    @Override
+    public void visit(LetExpressionNode letExpressionNode) {
+        addConstantReferencesDiagnostics(letExpressionNode.toSourceCode());
+    }
+
+    @Override
+    public void visit(ListConstructorExpressionNode listConstructorExpressionNode) {
+        listConstructorExpressionNode.expressions().forEach(exp -> visitSyntaxNode(exp));
+    }
+
+    @Override
+    public void visit(MappingConstructorExpressionNode mappingConstructorExpressionNode) {
+        mappingConstructorExpressionNode.fields().forEach(field -> visitSyntaxNode(field));
+    }
+
+    @Override
+    public void visit(MapTypeDescriptorNode mapTypeDescriptorNode) {
+        addConstantReferencesDiagnostics(mapTypeDescriptorNode.toSourceCode());
+    }
+
+    @Override
+    public void visit(MethodCallExpressionNode methodCallExpressionNode) {
+        visitSyntaxNode(methodCallExpressionNode.expression());
+        methodCallExpressionNode.arguments().forEach(arg -> visitSyntaxNode(arg));
+    }
+
+    @Override
+    public void visit(SimpleNameReferenceNode nameReferenceNode) {
+        addConstantReferencesDiagnostics(nameReferenceNode.toSourceCode());
+    }
+
+    @Override
+    public void visit(QualifiedNameReferenceNode nameReferenceNode) {
+        addConstantReferencesDiagnostics(nameReferenceNode.toSourceCode());
+    }
+
+    @Override
+    public void visit(NaturalExpressionNode naturalExpressionNode) {
+        addConstantReferencesDiagnostics(naturalExpressionNode.toSourceCode());
+    }
+
+    @Override
+    public void visit(NilLiteralNode nilLiteralNode) {
+        // Allowed in constant expressions.
+    }
+
+    @Override
+    public void visit(ObjectConstructorExpressionNode objectConstructorExpressionNode) {
+        addConstantReferencesDiagnostics(objectConstructorExpressionNode.toSourceCode());
+    }
+
+    @Override
+    public void visit(ObjectTypeDescriptorNode objectTypeDescriptorNode) {
+        addConstantReferencesDiagnostics(objectTypeDescriptorNode.toSourceCode());
+    }
+
+    @Override
+    public void visit(OptionalFieldAccessExpressionNode optionalFieldAccessExpressionNode) {
+        visitSyntaxNode(optionalFieldAccessExpressionNode.expression());
```

Comment:
I intentionally allow this to support field access with constant values.
cc @MaryamZi